packages/context/src/decorator-factory.ts

```diff
@@ -0,0 +1,269 @@
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Reflector} from './reflect';
+import * as _ from 'lodash';
+import * as debugModule from 'debug';
+const debug = debugModule('loopback:decorator');
+
+function cloneDeep<T>(val: T): T {
+  if (val === undefined) {
+    return {} as T;
+  }
+  return _.cloneDeepWith(val, v => {
+    // Do not clone functions
+    if (typeof v === 'function') return v;
+    return undefined;
+  });
+}
+
+/**
+ * An object mapping keys to corresponding metadata
+ */
+export interface MetadataMap<T> {
+  [property: string]: T;
+}
+
+/**
+ * Decorator function types
+ */
+export type DecoratorType =
+  | ClassDecorator
+  | PropertyDecorator
+  | MethodDecorator
+  | ParameterDecorator;
+
+// tslint:disable:no-any
+/**
+ * Base factory class for decorator functions
+ *
+ * @example
+ * ```
+ * function classDecorator(spec: MySpec): ClassDecorator {
+ *   return ClassDecoratorFactory.getDecorator('my-key', spec);
+ * }
+ * ```
+ * or
+ * ```
+ * function classDecorator(spec: MySpec): ClassDecorator {
+ *   const factory: ClassDecoratorFactory<MySpec>('my-key', spec);
+ *   return factory.create();
+ * }
+ * ```
+ * These functions above declare `@classDecorator` that can be used as follows:
+ * ```
+ * @classDecorator({x: 1})
+ * class MyController {}
+ * ```
+ */
+export class DecoratorFactory<T> {
+  /**
+   * Construct a new class decorator factory
+   * @param key Metadata key
+   * @param spec Metadata object from the decorator function
+   * @param allowInheritance Inheritance will be honored, default to `true`
+   */
+  constructor(
+    protected key: string,
+    protected spec: T,
+    protected allowInheritance = true,
+  ) {}
+
+  /**
+   * Inherit metadata from base classes. By default, this method merges base
+   * metadata into the spec if `allowInheritance` is set to `true`. To customize
+   * the behavior, this method can be overridden by sub classes.
+   *
+   * @param baseMeta Metadata from base classes
+   */
+  inherit(baseMeta: T | undefined | null): T {
+    if (!this.allowInheritance) return this.spec;
+    if (baseMeta == null) return this.spec;
+    if (this.spec == undefined) return baseMeta;
+    if (typeof baseMeta !== 'object') return this.spec;
+    if (Array.isArray(baseMeta) || Array.isArray(this.spec)) {
+      // For arrays, we don't merge
+      return this.spec;
+    }
+    return Object.assign(baseMeta, this.spec);
+  }
+
+  /**
+   * Create a decorator function of the given type
+   * Each some class should implement this method
+   */
+  create(): DecoratorType {
+    throw new Error('create() is not implemented');
+  }
+
+  /**
+   * Get the decorator function
+   * @param key Metadata key
+   * @param spec Metadata object from the decorator function
+   * @param allowInheritance Inheritance will be honored, default to `true`
+   */
+  static getDecorator<D extends DecoratorType>(
```

Comment:
`getDecorator` creates the impression that an existing decorator function is returned. Can we use `createDecorator` instead? I think that's the usual convention in Factory Design Pattern.